src/utils.jl

```diff
@@ -4,19 +4,19 @@ $(SIGNATURES)
 Returns the number of arguments of `f` for each method.
 """
 function numargs(f)
-    return [num_types_in_tuple(m.sig) - 1 for m in methods(f)] #-1 since f is the first parameter
+    return Tuple(num_types_in_tuple(m.sig) - 1 for m in Tricks.static_methods(f)) #-1 since f is the first parameter
 end
 
 """
 $(SIGNATURES)
 
 Get the number of parameters of a Tuple type, i.e. the number of fields.
 """
-function num_types_in_tuple(sig)
+Base.@pure function num_types_in_tuple(sig)
```

Comment:
I probably lose any sense of inference in here. It probably needs hand rolled generated functions?